libs/user/user.ts

```diff
@@ -0,0 +1,118 @@
+import Queue from "queue";
+import path from "path";
+import {
+  CONTRACT_SHIELD,
+  NIGHTFALL_DEFAULT_CONFIG,
+  TX_FEE_DEFAULT,
+} from "./constants";
+import { UserConfig, UserDeposit } from "./types";
+import {
+  getEthAddressFromPrivateKey,
+  createZkpKeysFromMnemonic,
+} from "../keys";
+import { Client } from "../client";
+import { parentLogger, Web3Websocket } from "../utils";
+import { createDeposit } from "../transactions/deposit";
+
+const logger = parentLogger.child({
+  name: path.relative(process.cwd(), __filename),
+});
+
+class User {
+  // constructor
+  blockchainNetwork: string;
+  blockchainWs: string;
+  apiUrl: string;
+  web3Websocket;
+  client;
+
+  // init
+  shieldContractAddress: null | string = null;
+  ethPrivateKey: null | string = null;
+  ethAddress: null | string = null;
+  nightfallMnemonic: null | string = null;
+  zkpKeys: any = null;
+
+  // when transacting
+  token: any = null;
+  txQueue: Queue = null;
+
+  constructor(env = NIGHTFALL_DEFAULT_CONFIG) {
+    logger.debug({ env }, "new User connected to");
+
+    this.blockchainNetwork = env.blockchainNetwork.toLowerCase();
+    this.blockchainWs = env.blockchainWs.toLowerCase();
+    this.apiUrl = env.apiUrl.toLowerCase();
+
+    this.web3Websocket = new Web3Websocket(this.blockchainWs);
+    this.client = new Client(this.apiUrl);
+  }
+
+  // TODO improve return type
+  async init(config: UserConfig) {
```

Comment:
Is there some problem if we don't expose this method and call it from the constructor? it isn't good leave on the client's hand the obligation of initializing an object.

Reply:
Hey @imagobea, I think that a solution for this is receiving on the constructor the parameters for which an `await` is otherwise required. Instead of someone instantiating a User directly, one will ask it to a `UserFactory` (Factory Method design pattern), which knows how to do it. It would be more or less like this...
```
class UserFactory {
    function createUser(env = NIGHTFALL_DEFAULT_CONFIG, config: UserConfig): User {
        const apiUrl = env.apiUrl.toLowerCase();
        const client = new Client(apiUrl);
        const shieldContractAddress = await client.getContractAddress(CONTRACT_SHIELD);
        const nightfallKeys = await createZkpKeysFromMnemonic(config.nightfallMnemonic, client);
            
        return new User(env, client, shieldContractAddress, nightfallKeys);
    }
}

...
class User {
    ...
    constructor(env = NIGHTFALL_DEFAULT_CONFIG, client, shieldContractAddress, nightfallKeys) {
        ...
        this.client = client;
        this.shieldContractAddress = shieldContractAddress;
        this.nightfallKeys = nightfallKeys;
    }
    ...
}
```